File: python/dkim2sign.py
#!/usr/bin/env python3
"""
DKIM2 signer - draft-ietf-dkim-dkim2-spec-01

Takes a raw email, selector, domain, and keyfile and produces a signed
message with Message-Instance and DKIM2-Signature headers on stdout.
"""

import argparse
import base64
import hashlib
import io
import json
import re
import sys
import time
from pathlib import Path
from typing import IO, Union

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec, ed25519, padding, rsa, utils


# Source can be raw bytes, a file path (str/Path), or any binary-mode file-like object.
Source = Union[bytes, str, "Path", "IO[bytes]"]


def _to_bytes(source: "Source") -> bytes:
    """Normalize any message source to bytes."""
    if isinstance(source, bytes):
        return source
    if isinstance(source, (str, Path)):
        return Path(source).read_bytes()
    return source.read()


# ---------------------------------------------------------------------------
# Utility helpers
# ---------------------------------------------------------------------------

def b64(data: bytes) -> str:
    """Base64-encode bytes, return str with no newlines."""
    return base64.b64encode(data).decode("ascii")


def b64json(obj) -> str:
    """JSON-encode an object, then base64-encode the result."""
    return b64(json.dumps(obj, separators=(",", ":")).encode("utf-8"))


# ---------------------------------------------------------------------------
# Parse raw message into (header_lines, body)
# ---------------------------------------------------------------------------

def parse_message(source: "Source") -> tuple[list[bytes], bytes]:
    """Split a raw RFC 5322 message into header lines and body.

    Returns (headers, body) where headers is a list of complete header
    fields (including continuation lines) as raw bytes, and body is the
    raw body bytes (everything after the blank line separator).
    """
    raw = _to_bytes(source)
    # Normalise line endings to CRLF
    raw = raw.replace(b"\r\n", b"\n").replace(b"\r", b"\n").replace(b"\n", b"\r\n")

    # Split at first blank line
    sep = b"\r\n\r\n"
    idx = raw.find(sep)
    if idx == -1:
        header_block = raw
        body = b""
    else:
        header_block = raw[:idx]
        body = raw[idx + len(sep):]

    # Split header block into individual header fields (handling continuations)
    header_lines: list[bytes] = []
    for line in header_block.split(b"\r\n"):
        if line and line[0:1] in (b" ", b"\t") and header_lines:
            # Continuation line – append to previous header
            header_lines[-1] += b"\r\n" + line
        else:
            if line:
                header_lines.append(line)

    return header_lines, body


# ---------------------------------------------------------------------------
# Canonicalization for header hash (Section 5.2)
# ---------------------------------------------------------------------------

# Headers to exclude from the header hash
_EXCLUDED_PREFIXES = (b"x-", b"arc-")
_EXCLUDED_NAMES = {b"received", b"return-path", b"delivered-to",
                   b"message-instance",
                   b"dkim2-signature", b"dkim-signature",
                   b"authentication-results"}


def _header_name(hdr: bytes) -> bytes:
    """Extract the field name from a raw header line (before the colon)."""
    colon = hdr.find(b":")
    if colon == -1:
        return hdr.strip().lower()
    return hdr[:colon].strip().lower()


def _should_exclude_header(name: bytes) -> bool:
    """Return True if this header should be excluded from the header hash."""
    if name in _EXCLUDED_NAMES:
        return True
    for prefix in _EXCLUDED_PREFIXES:
        if name.startswith(prefix):
            return True
    return False


def canonicalize_header_field(raw_hdr: bytes) -> bytes:
    """Canonicalize a single header field per Section 5.2 steps 2-6.

    Input is a raw header field (may include continuation lines).
    Returns the canonicalized form (name:value without trailing CRLF).
    """
    # Decode to str for easier manipulation
    hdr = raw_hdr.decode("utf-8", errors="surrogateescape")

    # Step 3: Unfold continuation lines (remove CRLF before WSP)
    hdr = re.sub(r"\r\n([ \t])", r"\1", hdr)

    # Split into name and value at first colon
    colon = hdr.find(":")
    if colon == -1:
        name = hdr
        value = ""
    else:
        name = hdr[:colon]
        value = hdr[colon + 1:]

    # Step 2: Lowercase the field name
    name = name.lower()

    # Step 4: Convert all WSP sequences to single SP
    name = re.sub(r"[ \t]+", " ", name).strip()
    value = re.sub(r"[ \t]+", " ", value)

    # Step 5: Delete WSP at end of value
    value = value.rstrip(" \t")

    # Step 6: Delete WSP before and after colon (i.e. trim name and value)
    name = name.strip()
    value = value.lstrip(" \t")

    return (name + ":" + value).encode("utf-8", errors="surrogateescape")


def compute_header_hash(headers: list[bytes]) -> bytes:
    """Compute the SHA-256 hash of canonicalized, sorted headers (Section 5.2).

    Excludes headers listed in the spec. Returns raw digest bytes.
    """
    canon_headers: list[tuple[bytes, bytes]] = []

    for hdr in headers:
        name = _header_name(hdr)
        if _should_exclude_header(name):
            continue
        canon = canonicalize_header_field(hdr)
        canon_headers.append((name, canon))

    # Step 7-8: Sort alphabetically by name; duplicate names are ordered
    # bottom-up (last occurrence first), matching recipe numbering.
    # Reverse before sorting so Python's stable sort preserves bottom-up order.
    canon_headers.reverse()
    canon_headers.sort(key=lambda x: x[0])

    # Concatenate with CRLF separators and a trailing CRLF
    data = b"\r\n".join(ch for _, ch in canon_headers)
    if data:
        data += b"\r\n"

    return hashlib.sha256(data).digest()


# ---------------------------------------------------------------------------
# Canonicalization for body hash (Section 5.1)
# ---------------------------------------------------------------------------

def compute_body_hash(body: bytes) -> bytes:
    """Compute the SHA-256 hash of the canonicalized body (Section 5.1).

    Simple canonicalization:
    - Strip all trailing empty lines
    - Ensure body ends with exactly one CRLF
    """
    # Normalise line endings
    body = body.replace(b"\r\n", b"\n").replace(b"\r", b"\n").replace(b"\n", b"\r\n")

    # Remove all trailing CRLF sequences (empty lines at end)
    while body.endswith(b"\r\n"):
        body = body[:-2]

    # Add exactly one trailing CRLF (even if body was empty)
    body += b"\r\n"

    return hashlib.sha256(body).digest()


# ---------------------------------------------------------------------------
# Build Message-Instance header
# ---------------------------------------------------------------------------

def build_message_instance(headers: list[bytes], body: bytes,
                           version: int = 1, recipe: dict | None = None) -> str:
    """Build a Message-Instance header field value.

    Returns the complete header as a string (including field name).
    Trailing semicolon is included per spec ABNF (tag-list grammar).
    """
    h_hash = compute_header_hash(headers)
    b_hash = compute_body_hash(body)
    value = f"m={version}; h=sha256:{b64(h_hash)}:{b64(b_hash)}"
    if recipe is not None:
        value += f"; r={b64json(recipe)}"
    value += ";"
    return f"Message-Instance: {value}"


# ---------------------------------------------------------------------------
# Signature computation (Section 11.5)
# ---------------------------------------------------------------------------

def canonicalize_sig_header(raw_hdr: str) -> bytes:
    """Canonicalize a Message-Instance or DKIM2-Signature header for signing.

    Per Section 9.5: same as header hash except ALL WSP is deleted
    (not collapsed to single SP).
    """
    hdr = raw_hdr if isinstance(raw_hdr, str) else raw_hdr.decode("utf-8", errors="surrogateescape")
    # Step 3: Unfold continuation lines
    hdr = re.sub(r"\r\n([ \t])", r"\1", hdr)
    # Split into name and value at first colon
    colon = hdr.find(":")
    if colon == -1:
        name = hdr
        value = ""
    else:
        name = hdr[:colon]
        value = hdr[colon + 1:]
    # Step 2: Lowercase the field name
    name = name.lower().strip()
    # Step 3 (sig-specific): Delete ALL WSP characters
    value = re.sub(r"[ \t]+", "", value)
    # Strip trailing CRLF/LF
    value = value.rstrip("\r\n")
    return (name + ":" + value + "\r\n").encode("utf-8", errors="surrogateescape")


def _extract_tag(header_value: str, tag: str) -> str | None:
    """Extract a tag value from a DKIM2-style tag-list header value."""
    # header_value is the part after "HeaderName: "
    for part in header_value.split(";"):
        part = part.strip()
        if part.startswith(tag + "="):
            return part[len(tag) + 1:].strip()
    return None


def _get_version_from_mi(hdr: str) -> int:
    """Extract m= value from a Message-Instance header string."""
    # hdr is "Message-Instance: m=N; ..."
    colon = hdr.find(":")
    value = hdr[colon + 1:] if colon != -1 else hdr
    m = _extract_tag(value, "m")
    return int(m) if m else 0


def _get_seq_from_sig(hdr: str) -> int:
    """Extract i= value from a DKIM2-Signature header string."""
    colon = hdr.find(":")
    value = hdr[colon + 1:] if colon != -1 else hdr
    v = _extract_tag(value, "i")
    return int(v) if v else 0


def compute_signature(mi_headers: list[str], sig_headers: list[str],
                      incomplete_sig: str, private_key, algorithm: str) -> bytes:
    """Compute the DKIM2 signature over MI and DKIM2-Sig headers.

    Args:
        mi_headers: Existing Message-Instance headers (as full header strings)
        sig_headers: Existing DKIM2-Signature headers (as full header strings)
        incomplete_sig: The DKIM2-Signature being created, with s= tag empty
        private_key: Private key object (RSA or Ed25519)
        algorithm: "rsa" or "ed25519"

    Returns:
        Raw signature bytes.
    """
    # Per draft-ietf-dkim-dkim2-spec-01 Section 9.5:
    # 1. All MI headers in ascending v= order
    # 2. All prior DKIM2-Signature headers in ascending i= order
    # 3. The incomplete DKIM2-Signature being created
    ordered: list[str] = []
    ordered.extend(sorted(mi_headers, key=_get_version_from_mi))
    ordered.extend(sorted(sig_headers, key=_get_seq_from_sig))
    ordered.append(incomplete_sig)

    # Canonicalize each header; each already ends in CRLF per spec Section 8.5
    canon = [canonicalize_sig_header(h) for h in ordered]
    data = b"".join(canon)

    # Hash with SHA-256
    digest = hashlib.sha256(data).digest()

    # Sign
    if algorithm.startswith("ed25519"):
        # Ed25519 signs the raw data (PureEdDSA), but the spec says
        # "signs the hash" - so we sign the SHA-256 digest
        return private_key.sign(digest)
    elif algorithm.startswith("rsa"):
        return private_key.sign(
            digest,
            padding.PKCS1v15(),
            utils.Prehashed(hashes.SHA256()),
        )
    else:
        raise ValueError(f"Unsupported algorithm: {algorithm}")


# ---------------------------------------------------------------------------
# Build DKIM2-Signature header
# ---------------------------------------------------------------------------

def build_dkim2_signature(mi_headers: list[str], sig_headers: list[str],
                          new_mi: str, domain: str, selector: str,
                          private_key, algorithm: str,
                          mailfrom: str = "<>",
                          rcptto: list[str] | None = None,
                          seq: int = 1, mi_version: int = 1,
                          timestamp: int | None = None,
                          next_domain: str | None = None,
                          flags: list[str] | None = None) -> str:
    """Build a complete DKIM2-Signature header.

    If next_domain is given, the signature carries an nd= tag for an imaginary
    forwarding hop (draft-03 §9.3) and omits mf=/rt=. Otherwise it carries
    mf=/rt= as usual. Any flags are emitted as an f= tag (draft-03 §8.10).

    Returns the full header string including field name.
    """
    if timestamp is None:
        timestamp = int(time.time())

    # draft-03 §9.3: an nd= hop carries nd= instead of mf=/rt=.
    if next_domain:
        chain = f"nd={next_domain}"
    else:
        mf_b64 = b64(mailfrom.encode("utf-8"))
        rt_list = rcptto or ["unknown@example.com"]
        rt_b64 = ",".join(b64(r.encode("utf-8")) for r in rt_list)
        chain = f"mf={mf_b64}; rt={rt_b64}"

    f_tag = f" f={','.join(flags)};" if flags else ""

    # Build the incomplete signature header with sel:alg: (null/empty string per spec §9.6).
    # Trailing semicolon included per spec ABNF (tag-list grammar).
    incomplete = (
        f"DKIM2-Signature: i={seq}; m={mi_version}; t={timestamp}; "
        f"d={domain}; {chain}; s={selector}:{algorithm}:;{f_tag}"
    )

    # Collect all MI headers including the new one
    all_mi = mi_headers + [new_mi]

    # Compute signature
    sig_bytes = compute_signature(all_mi, sig_headers, incomplete,
                                 private_key, algorithm)

    # Build s= tag: sel:alg:sig
    s_complete = f"{selector}:{algorithm}:{b64(sig_bytes)}"

    # Build the final header with the actual signature value
    return (
        f"DKIM2-Signature: i={seq}; m={mi_version}; t={timestamp}; "
        f"d={domain}; {chain}; s={s_complete};{f_tag}"
    )


# ---------------------------------------------------------------------------
# Key loading
# ---------------------------------------------------------------------------

def load_private_key(keyfile: str) -> tuple:
    """Load a private key from a PEM file.

    Returns (key_object, algorithm_string).
    """
    key_data = Path(keyfile).read_bytes()

    key = serialization.load_pem_private_key(key_data, password=None)

    if isinstance(key, ed25519.Ed25519PrivateKey):
        return key, "ed25519-sha256"
    elif isinstance(key, rsa.RSAPrivateKey):
        return key, "rsa-sha256"
    else:
        raise ValueError(f"Unsupported key type: {type(key)}")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def sign_message(source: "Source", selector: str, domain: str, keyfile: str,
                 mailfrom: str = "<>", rcptto: list[str] | None = None,
                 timestamp: int | None = None) -> bytes:
    """Sign a raw email message with DKIM2.

    Returns the complete message with Message-Instance and DKIM2-Signature
    headers prepended.
    """
    raw = _to_bytes(source)
    private_key, algorithm = load_private_key(keyfile)

    headers, body = parse_message(raw)

    # Find any existing Message-Instance and DKIM2-Signature headers
    existing_mi: list[str] = []
    existing_sig: list[str] = []
    for hdr in headers:
        name = _header_name(hdr)
        if name == b"message-instance":
            existing_mi.append(hdr.decode("utf-8", errors="surrogateescape"))
        elif name == b"dkim2-signature":
            existing_sig.append(hdr.decode("utf-8", errors="surrogateescape"))

    # Determine version numbers
    mi_version = 1
    if existing_mi:
        mi_version = max(_get_version_from_mi(h) for h in existing_mi) + 1

    sig_seq = 1
    if existing_sig:
        sig_seq = max(_get_seq_from_sig(h) for h in existing_sig) + 1

    # Build Message-Instance header
    mi_hdr = build_message_instance(headers, body, version=mi_version)

    # Build DKIM2-Signature header
    sig_hdr = build_dkim2_signature(
        existing_mi, existing_sig, mi_hdr,
        domain, selector, private_key, algorithm,
        mailfrom=mailfrom, rcptto=rcptto,
        seq=sig_seq, mi_version=mi_version,
        timestamp=timestamp,
    )

    # Reassemble the message with new headers prepended
    # Normalise line endings in original
    raw = raw.replace(b"\r\n", b"\n").replace(b"\r", b"\n").replace(b"\n", b"\r\n")

    output = sig_hdr.encode("utf-8") + b"\r\n"
    output += mi_hdr.encode("utf-8") + b"\r\n"
    output += raw

    return output


def main():
    parser = argparse.ArgumentParser(
        description="Sign an email with DKIM2 (draft-ietf-dkim-dkim2-spec-01)")
    parser.add_argument("message", help="Path to raw email file (- for stdin)")
    parser.add_argument("-s", "--selector", required=True,
                        help="DKIM2 selector name")
    parser.add_argument("-d", "--domain", required=True,
                        help="Signing domain")
    parser.add_argument("-k", "--keyfile", required=True,
                        help="Path to PEM private key file")
    parser.add_argument("--mailfrom", default="<>",
                        help="MAIL FROM value (default: <>)")
    parser.add_argument("--rcptto", action="append",
                        help="RCPT TO value(s) (repeatable)")
    parser.add_argument("--timestamp", type=int, default=None,
                        help="Unix timestamp (default: current time)")
    args = parser.parse_args()

    if args.message == "-":
        raw = sys.stdin.buffer.read()
    else:
        raw = Path(args.message).read_bytes()

    rcptto = args.rcptto or ["unknown@example.com"]

    result = sign_message(raw, args.selector, args.domain, args.keyfile,
                          mailfrom=args.mailfrom, rcptto=rcptto,
                          timestamp=args.timestamp)

    sys.stdout.buffer.write(result)


if __name__ == "__main__":
    main()

File: c/dkim2_sign.c
#include "dkim2_sign.h"
#include "dkim2_hash.h"
#include "dkim2_header.h"
#include "dkim2_crypto.h"
#include "base64.h"
#include <time.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <ctype.h>
#include <openssl/evp.h>

/* RFC5321 path for mf=/rt= (spec 7.5/7.6): brackets MUST be present.
 * Returns a malloc'd string the caller frees. NULL/empty -> "<>". */
static char *to_rfc5321_path(const char *addr) {
    if (!addr || !*addr) return strdup("<>");
    size_t n = strlen(addr);
    if (addr[0] == '<' && addr[n-1] == '>') return strdup(addr);
    char *out = malloc(n + 3);
    snprintf(out, n + 3, "<%s>", addr);
    return out;
}

/* §8.5 canonicalization for signature input:
   lowercase name, unfold (remove CRLF+following WSP), delete ALL remaining WSP,
   re-add trailing CRLF.
   Input: "Name: folded value\r\n"
   Returns malloc'd string or NULL. */
static char *canon_for_sig(const char *hdr) {
    size_t len = strlen(hdr);
    char *out = malloc(len + 4);
    if (!out) return NULL;
    char *p = out;

    const char *h = hdr;
    const char *colon = strchr(h, ':');
    if (!colon) { free(out); return NULL; }

    /* Lowercase name */
    for (; h < colon; h++) *p++ = (char)tolower((unsigned char)*h);
    *p++ = ':';
    h++; /* skip colon */

    /* Process value: unfold, delete ALL whitespace */
    while (*h) {
        if (h[0] == '\r' && h[1] == '\n') {
            h += 2;
            while (*h == ' ' || *h == '\t') h++; /* skip fold WSP */
            continue;
        }
        if (*h == ' ' || *h == '\t') { h++; continue; }
        *p++ = *h++;
    }

    /* Ensure trailing CRLF */
    if (p > out && p[-1] == '\n') {
        /* Already ends with \n — check for \r */
        if (p > out + 1 && p[-2] == '\r') {
            /* already \r\n */
        } else {
            /* just \n — replace with \r\n */
            p[-1] = '\r'; *p++ = '\n';
        }
    } else {
        *p++ = '\r'; *p++ = '\n';
    }
    *p = '\0';
    return out;
}

/* Build the §8.5 signing input:
   Canonicalized MI headers (ascending m=), then DKIM2-Signature headers
   (ascending i=), then the incomplete new signature header (empty sig values).
   Returns malloc'd buffer and sets *out_len. */
static unsigned char *build_sign_input(
    dkim2_mi_t *mi_list, dkim2_sig_t *sig_list,
    const char *new_sig_value, size_t *out_len) {
    char *buf = malloc(65536);
    if (!buf) return NULL;
    size_t pos = 0;

    /* MI headers ascending m= — use raw_value when available (existing MIs),
       else format from struct (newly created MI with no raw_value yet) */
    for (dkim2_mi_t *mi = mi_list; mi; mi = mi->next) {
        char *mi_val = mi->raw_value ? NULL : dkim2_mi_format(mi);
        const char *use_val = mi->raw_value ? mi->raw_value : mi_val;
        if (!use_val) { free(buf); return NULL; }
        size_t vl = strlen(use_val);
        char *full = malloc(vl + 24);
        snprintf(full, vl + 24, "message-instance: %s\r\n", use_val);
        free(mi_val);
        char *canon = canon_for_sig(full);
        free(full);
        if (canon) {
            size_t cl = strlen(canon);
            if (pos + cl < 65536) { memcpy(buf + pos, canon, cl); pos += cl; }
            free(canon);
        }
    }

    /* Existing DKIM2-Signature headers ascending i= */
    for (dkim2_sig_t *sig = sig_list; sig; sig = sig->next) {
        /* Use raw_value which is the complete original header value */
        size_t rv_len = strlen(sig->raw_value);
        char *full = malloc(rv_len + 24);
        snprintf(full, rv_len + 24, "dkim2-signature: %s\r\n", sig->raw_value);
        char *canon = canon_for_sig(full);
        free(full);
        if (canon) {
            size_t cl = strlen(canon);
            if (pos + cl < 65536) { memcpy(buf + pos, canon, cl); pos += cl; }
            free(canon);
        }
    }

    /* Incomplete new DKIM2-Signature (empty sig values) */
    size_t ns_len = strlen(new_sig_value);
    char *full = malloc(ns_len + 24);
    snprintf(full, ns_len + 24, "dkim2-signature: %s\r\n", new_sig_value);
    char *canon = canon_for_sig(full);
    free(full);
    if (canon) {
        size_t cl = strlen(canon);
        if (pos + cl < 65536) { memcpy(buf + pos, canon, cl); pos += cl; }
        free(canon);
    }

    *out_len = pos;
    return (unsigned char *)buf;
}

int dkim2_do_sign(dkim2_ctx_t *ctx, const dkim2_sign_config_t *cfg,
    char **mi_out, char **sig_out) {
    /* spec-05 §3.1: which hash algorithm(s) to emit in h=. Default (cfg->hash
       NULL) is sha256 only, so default output stays byte-identical to before
       hash agility existed. "--hash both" emits sha256 first, then sha512. */
    int sel_algs[DKIM2_N_HASH_ALGS];
    int n_sel = 0;
    const char *want = cfg->hash ? cfg->hash : "sha256";
    if (strcmp(want, "both") == 0) {
        sel_algs[n_sel++] = 0;
        sel_algs[n_sel++] = 1;
    } else {
        int a = dkim2_hash_alg_index(want);
        if (a < 0) a = 0;
        sel_algs[n_sel++] = a;
    }

    /* §8.3/§8.4: body hash already computed incrementally (both algorithms);
       header hash computed here per selected algorithm. */
    dkim2_hashset_t hs[DKIM2_N_HASH_ALGS];
    char hh_b64[DKIM2_N_HASH_ALGS][DKIM2_MAX_HASH_LEN * 2];
    char bh_b64[DKIM2_N_HASH_ALGS][DKIM2_MAX_HASH_LEN * 2];
    for (int k = 0; k < n_sel; k++) {
        int a = sel_algs[k];
        size_t alen = dkim2_hash_alg_len(a);

        unsigned char hd[DKIM2_MAX_HASH_LEN];
        if (dkim2_header_hash_raw_alg((const char **)ctx->headers, ctx->n_headers, a, hd) < 0) {
            snprintf(ctx->errmsg, sizeof ctx->errmsg, "header hash failed");
            return -1;
        }
        if (b64_encode(hd, alen, hh_b64[k], sizeof hh_b64[k]) < 0) {
            snprintf(ctx->errmsg, sizeof ctx->errmsg, "header hash encode failed");
            return -1;
        }
        if (b64_encode(ctx->body_digests.d[a], alen, bh_b64[k], sizeof bh_b64[k]) < 0) {
            snprintf(ctx->errmsg, sizeof ctx->errmsg, "body hash encode failed");
            return -1;
        }
        hs[k].alg       = (char *)dkim2_hash_alg_name(a);
        hs[k].hdr_hash  = hh_b64[k];
        hs[k].body_hash = bh_b64[k];
    }

    /* Determine m= for new Message-Instance */
    int new_m = 1;
    for (dkim2_mi_t *mi = ctx->mi_list; mi; mi = mi->next)
        if (mi->m >= new_m) new_m = mi->m + 1;

    /* Check if highest existing MI already has these exact hashes. Only
       applies to the legacy single-sha256 path so --hash semantics stay
       simple: a multi-hash sign always adds a new MI. */
    dkim2_mi_t *latest_mi = NULL;
    for (dkim2_mi_t *mi = ctx->mi_list; mi; mi = mi->next) latest_mi = mi;
    if (n_sel == 1 && sel_algs[0] == 0 &&
        latest_mi && latest_mi->n_hsets > 0 &&
        strcmp(latest_mi->hsets[0].hdr_hash, hh_b64[0]) == 0 &&
        strcmp(latest_mi->hsets[0].body_hash, bh_b64[0]) == 0) {
        /* Reuse existing MI — sign against it */
        new_m = latest_mi->m;
    }

    /* Build the new MI struct for signing input */
    dkim2_mi_t new_mi = {0};
    new_mi.m = new_m;
    new_mi.hsets = hs;
    new_mi.n_hsets = n_sel;
    new_mi.next = NULL;

    /* Determine i= for new DKIM2-Signature */
    int new_i = 1;
    for (dkim2_sig_t *sig = ctx->sig_list; sig; sig = sig->next)
        if (sig->i >= new_i) new_i = sig->i + 1;

    /* Base64-encode mf= — normalized to a bracketed RFC5321 path (spec §7.5) */
    char *mfp = to_rfc5321_path(ctx->mail_from);
    size_t mf_b64_len = ((strlen(mfp) + 2) / 3) * 4 + 2;
    char *mf_b64 = malloc(mf_b64_len);
    b64_encode((const unsigned char *)mfp, strlen(mfp), mf_b64, mf_b64_len);
    free(mfp);

    /* Base64-encode each rt= value, comma-separated — each normalized to a
       bracketed RFC5321 path (spec §7.6) */
    /* Estimate size: each rcpt ~64 chars encoded */
    size_t rt_buf_size = 4096;
    char *rt_b64 = calloc(1, rt_buf_size);
    size_t rt_pos = 0;
    if (ctx->rcpt_to) {
        for (int i = 0; ctx->rcpt_to[i]; i++) {
            char *rcpt = to_rfc5321_path(ctx->rcpt_to[i]);
            size_t rcpt_enc_len = ((strlen(rcpt) + 2) / 3) * 4 + 2;
            char *enc = malloc(rcpt_enc_len);
            b64_encode((const unsigned char *)rcpt, strlen(rcpt), enc, rcpt_enc_len);
            free(rcpt);
            if (i) { rt_b64[rt_pos++] = ','; }
            size_t el = strlen(enc);
            memcpy(rt_b64 + rt_pos, enc, el);
            rt_pos += el;
            free(enc);
        }
    }
    rt_b64[rt_pos] = '\0';

    /* Load private key */
    EVP_PKEY *privkey = dkim2_load_privkey(cfg->privkey_path);
    if (!privkey) {
        snprintf(ctx->errmsg, sizeof ctx->errmsg, "failed to load private key: %s", cfg->privkey_path);
        free(mf_b64); free(rt_b64); return -1;
    }

    /* Auto-detect algorithm from key type if not specified */
    const char *alg = cfg->alg;
    if (!alg) {
        int key_id = EVP_PKEY_id(privkey);
        if (key_id == EVP_PKEY_ED25519)      alg = "ed25519-sha256";
        else if (key_id == EVP_PKEY_RSA)     alg = "rsa-sha256";
        else {
            EVP_PKEY_free(privkey);
            free(mf_b64); free(rt_b64);
            snprintf(ctx->errmsg, sizeof ctx->errmsg, "unsupported key type");
            return -1;
        }
    }

    /* Use configured timestamp or current time */
    uint64_t now = cfg->timestamp ? cfg->timestamp : (uint64_t)time(NULL);

    /* Format incomplete DKIM2-Signature value (empty sig per §8.5) */
    char incomplete_sig[4096];
    snprintf(incomplete_sig, sizeof incomplete_sig,
        "i=%d; m=%d; t=%llu; d=%s; mf=%s; rt=%s; s=%s:%s:;",
        new_i, new_m, (unsigned long long)now,
        cfg->domain, mf_b64, rt_b64,
        cfg->selector, alg);

    /* Build MI list to sign: existing + new (if new_m > any existing) */
    dkim2_mi_t *mi_for_sign = ctx->mi_list;
    dkim2_mi_t *last_existing = NULL;
    int already_in_list = 0;
    for (dkim2_mi_t *m = ctx->mi_list; m; m = m->next) {
        if (m->m == new_m) { already_in_list = 1; break; }
        last_existing = m;
    }
    if (!already_in_list) {
        /* Append new_mi at appropriate position */
        if (!mi_for_sign) {
            mi_for_sign = &new_mi;
        } else {
            /* Append at end */
            last_existing = ctx->mi_list;
            while (last_existing->next) last_existing = last_existing->next;
            last_existing->next = &new_mi;
        }
    }

    /* Build signing input */
    size_t sign_input_len;
    unsigned char *sign_input = build_sign_input(
        mi_for_sign, ctx->sig_list, incomplete_sig, &sign_input_len);

    /* Detach new_mi from list to avoid dangling pointer */
    if (!already_in_list) {
        if (mi_for_sign == &new_mi) {
            mi_for_sign = NULL;
        } else {
            for (dkim2_mi_t *m = ctx->mi_list; m; m = m->next) {
                if (m->next == &new_mi) { m->next = NULL; break; }
            }
        }
    }

    if (!sign_input) {
        EVP_PKEY_free(privkey);
        free(mf_b64); free(rt_b64);
        snprintf(ctx->errmsg, sizeof ctx->errmsg, "failed to build signing input");
        return -1;
    }

    /* Sign */
    char *sig_b64 = dkim2_sign(privkey, alg, sign_input, sign_input_len);
    EVP_PKEY_free(privkey);
    free(sign_input);

    if (!sig_b64) {
        free(mf_b64); free(rt_b64);
        snprintf(ctx->errmsg, sizeof ctx->errmsg, "signing failed");
        return -1;
    }

    /* Format complete DKIM2-Signature value */
    char complete_sig[4096];
    snprintf(complete_sig, sizeof complete_sig,
        "i=%d; m=%d; t=%llu; d=%s; mf=%s; rt=%s; s=%s:%s:%s;",
        new_i, new_m, (unsigned long long)now,
        cfg->domain, mf_b64, rt_b64,
        cfg->selector, alg, sig_b64);
    free(sig_b64);

    /* Format MI value (only if new MI was created). dkim2_mi_format joins
       every hash-set with commas, e.g. "sha256:hh:bh,sha512:hh:bh" for
       --hash both — deterministic sha256-then-sha512 order. */
    if (!already_in_list) {
        *mi_out = dkim2_mi_format(&new_mi);
    } else {
        *mi_out = NULL; /* no new MI header needed */
    }

    *sig_out = strdup(complete_sig);
    free(mf_b64); free(rt_b64);
    return 0;
}
